khronicle-android-lint/src/main/kotlin/WrongLogUsageDetector.kt

```diff
@@ -0,0 +1,138 @@
+        val evaluator = context.evaluator
+
+        if (evaluator.isMemberInClass(method, "android.util.Log")) {
+            context.report(
+                Incident(
+                    issue = ISSUE_LOG,
+                    scope = node,
+                    location = context.getLocation(node),
+                    message = "Using 'android.util.Log' instead of 'com.juul.khronicle.Log'",
+                    fix = quickFixIssueLog(node),
+                ),
+            )
+            return
+        }
+    }
+
+    private fun quickFixIssueLog(logCall: UCallExpression): LintFix {
+        val arguments = logCall.valueArguments
+        val methodName = logCall.methodName
+
+        val tag = arguments[0].asSourceString()
+        val message: String
+        val throwable: String?
+
+        when (arguments.size) {
+            2 -> {
+                val msgOrThrowable = arguments[1]
+                val isString = msgOrThrowable.getExpressionType()?.isString() ?: true
+
+                message = when {
+                    isString -> msgOrThrowable.asSourceString()
+                    else -> ""
+                }
+
+                throwable = when {
+                    isString -> null
+                    else -> msgOrThrowable.sourcePsi?.text
+                }
+            }
+
+            3 -> {
+                message = arguments[1].asSourceString()
+                throwable = arguments[2].sourcePsi?.text
+            }
+
+            else -> {
+                throw IllegalStateException("android.util.Log overloads should have 2 or 3 arguments")
+            }
+        }
+
+        val khronicleLogMethodName = androidLogMethodToKhronicleLogMethod[methodName]
+
+        val fixSource1 = buildString {
+            append("com.juul.khronicle.Log.$khronicleLogMethodName(tag = $tag")
+            if (throwable == null) {
+                append(")")
+            } else {
+                append(", throwable = $throwable)")
+            }
+
+            append(" { $message }")
+        }
+
+        val fixSource2 = buildString {
+            append("com.juul.khronicle.Log.$khronicleLogMethodName")
+            if (throwable != null) {
+                append("(throwable = $throwable)")
+            }
+
+            append(" { $message }")
+        }
+
+        val logCallSource = logCall.uastParent!!.sourcePsi?.text
+        return fix()
+            .group()
+            .add(
+                fix()
+                    .replace()
+                    .text(logCallSource)
+                    .shortenNames()
+                    .reformat(true)
+                    .with(fixSource1)
+                    .build(),
+            ).add(
+                fix()
+                    .replace()
+                    .text(logCallSource)
+                    .shortenNames()
+                    .reformat(true)
+                    .with(fixSource2)
+                    .build(),
+            ).build()
```

Comment:
@eygraber can you help me understand what this is doing? Documentation for ktlint in this regard (`group`) is sparse. ChatGPT makes it sound like this will attempt to apply both `fix`es consecutively; if that is the case: won't the first (with `tag`) be applied and the 2nd `fix` never be applied (since both fixes are matching the same `logCallSource` text)?

Reply:
This is for `lint`, not `ktlint`. What this will do is offer two quick fixes to the user in the IDE, one that includes the tag, and the other that doesn't.

Reply:
Ohh, explains my confusion. Makes sense. Thanks for the clarification.